### ISSUE TITLE
General undo/redo with default key bindings

### ISSUE THREAD

**Issue by clehene** [ahrm/sioyek#1011]
I understand that there may me an option to undo highlights. I read https://github.com/ahrm/sioyek/issues/633 however it doesn't seem to work, not documented, not user-friendly and completely breaks any flow which, presumably, is a major goal of Sioyek.

However undo/redo is one of the few standard usabity features. Not having them should be considered a major usabilty bug.

Moreover it needs to be universal, not just for highlights. 
Standard key bindings should be supported. That is if vim u/ctr-r or, universally, cmd-z cmd-shift-z

**Comment by ahrm** [comment]
Hmmm, what is your exact workflow that requires an undo? Note that I am not dismissing your problem, just I personally have never felt the need for an undo command and your workflow is clearly different from mine. Knowing the exact workflow that requires this is useful for me because it may allow me to implement a better or more efficient solution.

**Comment by clehene** [comment]
@ahrm -  For PDFs it's annotations. Note - I'm just starting and I don't know if Sioyek supports any other annotaions such as scribbles or comments as with Arobat. 
However I highlight using multiple colors (e.g. definitions are one color, highlights are in 3 colors/categories, etc)
The other part of any editing flow is that if I incorrectly change (annotate) something, I want to undo the last operation, regardless what it was without thinking about it. Hence expecting _undo_.

Similarly for delete, I want to delete an object regardless what it is. If I select a highlight, I should just be able to "d" - "dh" is redundant since I'm already on a highlight

**Comment by JonasFuksa** [comment]
I second this. With the new pre-release I am using freetext_bookmarks a lot to annotate papers. However, it happens sometimes that I accidentally delete a bookmark -- for example when I have selected a different one than I wanted, or when I delete a bookmark and then want to take it back. Similarly with freehand drawings, although I don't use those much. 

I understand that, as you say in #633 this was not such a big deal previously (loosing a highlight is not that big a deal), but with text annotations you can really lose something valuable.

**Comment by jcpayne** [comment]
I agree with the others: I would love some kind of "Undo".  For me, the usual need is when I highlight something and then want to delete it or change the color.  In other pdf readers, you can right-click on any part of a highlight and make those changes, but in Sioyek you have to re-highlight the text, which is a pain, and then invoke the dh command (unless I've missed something -- I'm a new user).

**Comment by jcpayne** [comment]
Thanks, that's useful.  But I still feel  that it would be so useful to be able to say "Oops, i didn't mean to highlight that bit..." and undo it.

**Comment by N-Demir** [comment]
An undo would be faster. +1 to the idea

**Comment by jcpayne** [comment]
I just realized that there is a subtle UI behavior going on here that I missed until now.  I'm adding the explanation here to help others who have the same issue.  

Assume that I have highlighted some text, and I want to remove the highlight:
1. Clicking on any text that is not highlighted and then typing "dh" (delete highlight) has no effect.
2. Re-selecting the highlight text and typing "dh" **_also_** has no effect -- this is what I had been trying.
3. Clicking anywhere inside the highlighted text -- but _not_ re-selecting it -- and then typing "dh" removes the highlight.  Easy, once you know the trick.
